src/core/middleware.js

```diff
@@ -21,42 +22,70 @@ export function prefixMiddleWare(req, res, next, { _config = config } = {}) {
   // Get language from URL or fall-back to detecting it from accept-language header.
   const acceptLanguage = req.headers['accept-language'];
   const { lang, isLangFromHeader } = getLanguage({ lang: langFromURL, acceptLanguage });
+  // Get the application from the UA if one wasn't specified in the URL (or
+  // if it turns out to be invalid).
+  const application = getClientApp(req.headers['user-agent']);
 
   const hasValidLang = isValidLang(langFromURL);
+  const hasValidLocaleException = isValidLocaleUrlException(appFromURL, { _config });
   const hasValidClientApp = isValidClientApp(appFromURL, { _config });
-  let hasValidUrlException = isValidUrlException(appFromURL, { _config });
+  let hasValidClientAppUrlException = isValidClientAppUrlException(appFromURL, { _config });
 
+  let isApplicationFromHeader = false;
+  let prependedOrMovedApplication = false;
   let prependedLang = false;
 
-  if (
+  if (hasValidLocaleException) {
+    log.info(dedent`Second part of URL is a locale exception (${URLParts[1]});
+      make sure the clientApp is valid`);
+
+    // Normally we look for a clientApp in the second part of a URL, but URLs
+    // that match a locale exception don't have a locale so we look for the
+    // clientApp in the first part of the URL.
+    if (!isValidClientApp(langFromURL, { _config })) {
+      URLParts[0] = application;
+      isApplicationFromHeader = true;
+      prependedOrMovedApplication = true;
+    }
+  } else if (
     (hasValidLang && langFromURL !== lang) || hasValidClientApp ||
-    hasValidUrlException
+    hasValidClientAppUrlException
   ) {
     // Replace the first part of the URL if:
     // * It's valid and we've mapped it e.g: pt -> pt-PT.
     // * The lang is invalid but we have a valid application
     //   e.g. /bogus/firefox/.
     log.info(`Replacing lang in URL ${URLParts[0]} -> ${lang}`);
     URLParts[0] = lang;
+  } else if (isValidLocaleUrlException(URLParts[0], { _config })) {
+    log.info(`Prepending clientApp to URL: ${application}`);
+    URLParts.splice(0, 0, application);
+    isApplicationFromHeader = true;
+    prependedOrMovedApplication = true;
   } else if (!hasValidLang) {
     // If lang wasn't valid or was missing prepend one.
     log.info(`Prepending lang to URL: ${lang}`);
     URLParts.splice(0, 0, lang);
     prependedLang = true;
     // If we've prepended the lang to the URL we need to re-check our
     // URL exception and make sure it's valid.
-    hasValidUrlException = isValidUrlException(URLParts[1], { _config });
+    hasValidClientAppUrlException = isValidClientAppUrlException(URLParts[1], { _config });
```

Comment:
I know we don't force 80 char lines but I would be grateful if this was under 80 because I like my split screen :)